client/src/main/scala/com.crobox.clickhouse/internal/ClickHouseExecutor.scala

```diff
@@ -55,4 +76,70 @@ private[clickhouse] trait ClickHouseExecutor extends LazyLogging {
     }
   }
 
+  protected def executeRequestInternal(host: Future[Uri],
+                                       query: String,
+                                       settings: QuerySettings,
+                                       entity: Option[RequestEntity] = None,
+                                       progressQueue: Option[SourceQueueWithComplete[QueryProgress]]): Future[String] =
+    host.flatMap(actualHost => {
+      val request = toRequest(actualHost, query, settings.withFallback(config), entity)
+      processClickhouseResponse(singleRequest(request), query, actualHost, progressQueue)
+    })
+
+  private def executeWithRetries(retries: Int, progressQueue: Option[SourceQueueWithComplete[QueryProgress]])(
+      request: () => Future[String]
+  ): Future[String] =
+    request().recoverWith {
+      // The http server closed the connection unexpectedly before delivering responses for 1 outstanding requests
+      case e: StreamTcpException if retries > 0 =>
+        progressQueue.foreach(_.offer(QueryRetry(e, (queryRetries - retries) + 1)))
+        logger.warn(s"Stream exception, retries left: $retries", e)
+        executeWithRetries(retries - 1, progressQueue)(request)
+      case e: RuntimeException
+          if e.getMessage.contains("The http server closed the connection unexpectedly") && retries > 0 =>
+        logger.warn(s"Unexpected connection closure, retries left: $retries", e)
+        progressQueue.foreach(_.offer(QueryRetry(e, (queryRetries - retries) + 1)))
+        executeWithRetries(retries - 1, progressQueue)(request)
+    }
+}
+
+object ClickHouseExecutor {
+  sealed trait QueryProgress
+  case object QueryAccepted                                 extends QueryProgress
+  case object QueryFinished                                 extends QueryProgress
+  case object QueryRejected                                 extends QueryProgress
+  case class QueryFailed(cause: Throwable)                  extends QueryProgress
+  case class QueryRetry(cause: Throwable, retryNumber: Int) extends QueryProgress
+
+  case class QuerySettings(readOnly: Option[Int] = None,
```

Comment:
Maybe we should make the readOnly flag more comprehensable, e.g. what is 0, 1 and 2?